File: src/c/resource_cache/default/src/resource_cache.c
// ke_resource_cache impl — kernel built-in. Open-addressed table with
// tombstones for both the handle→entry and path→handle directions.
//
// destroy_fn is per-cache (set at construction). All resources in this cache
// share the same destructor (because per-domain caches are owned by the
// subsystem that knows how to free them).

#include <kernel_engine/resource_cache/resource_cache.h>
#include <kernel_engine/allocator/allocator.h>

#include <stdalign.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

static uint64_t ke_hash_string(const char *s)
{
    if (!s) return 0;
    uint64_t h = 5381;
    int c;
    while ((c = (unsigned char)*s++)) h = h * 33 ^ (uint64_t)c;
    return h;
}

// ── Internal state ──────────────────────────────────────────────────────────
//
// Two parallel tables:
//   (a) `resources` — open-addressed table of registered resources, keyed by
//       handle. Holds the refcount. Tombstones tracked so `release` can
//       reclaim slots without breaking probe chains.
//   (b) `paths` — same shape, keyed by FNV hash of the path string. Maps
//       path → handle for dedup lookups. Path entries don't hold a refcount;
//       the resource entry in (a) does.

typedef struct slot {
    uint64_t key;       // 0 = empty, UINT64_MAX = tombstone
    uint32_t refcount;  // resource entries only
    uint32_t handle;    // resource entries: redundant; path entries: target handle
} slot;

#define RC_EMPTY     0ULL
#define RC_TOMBSTONE UINT64_MAX

typedef struct table {
    slot         *slots;
    size_t        capacity;  // power of two
    size_t        occupied;  // active + tombstones
    size_t        active;    // active only
    ke_allocator *allocator;
} table;

typedef struct rc_state {
    ke_resource_cache         api;
    ke_allocator             *allocator;
    ke_resource_destroy_func  destroy_fn;
    void                     *destroy_ctx;
    table                     resources;
    table                     paths;
} rc_state;

// ── Table ops ───────────────────────────────────────────────────────────────

static ke_result table_init(table *t, ke_allocator *alloc, size_t initial_capacity) {
    t->allocator = alloc;
    t->capacity  = initial_capacity;
    t->occupied  = 0;
    t->active    = 0;
    t->slots = (slot *)alloc->alloc(alloc, initial_capacity * sizeof(slot), alignof(slot));
    if (!t->slots) return KE_ERROR_OUT_OF_MEMORY;
    memset(t->slots, 0, initial_capacity * sizeof(slot));
    return KE_OK;
}

static void table_destroy(table *t) {
    if (t->slots && t->allocator) t->allocator->free(t->allocator, t->slots);
    t->slots = NULL;
    t->capacity = t->occupied = t->active = 0;
}

// Finds the slot for `key`. Returns the matching active entry, or the first
// usable slot (preferring tombstones over empty so reinsertion keeps probe
// chains short). `*out_found` discriminates.
static size_t table_probe(const table *t, uint64_t key, bool *out_found) {
    size_t mask  = t->capacity - 1;
    size_t index = (size_t)(key & mask);
    size_t first_tombstone = SIZE_MAX;
    for (size_t i = 0; i < t->capacity; i++) {
        const slot *s = &t->slots[index];
        if (s->key == RC_EMPTY) {
            *out_found = false;
            return (first_tombstone != SIZE_MAX) ? first_tombstone : index;
        }
        if (s->key == RC_TOMBSTONE) {
            if (first_tombstone == SIZE_MAX) first_tombstone = index;
        } else if (s->key == key) {
            *out_found = true;
            return index;
        }
        index = (index + 1) & mask;
    }
    *out_found = false;
    return (first_tombstone != SIZE_MAX) ? first_tombstone : 0;
}

static ke_result table_rehash(table *t, size_t new_capacity);

static ke_result table_reserve(table *t) {
    if ((t->occupied + 1) * 10 < t->capacity * 7) return KE_OK;
    return table_rehash(t, t->capacity * 2);
}

static ke_result table_rehash(table *t, size_t new_capacity) {
    slot *old_slots = t->slots;
    size_t old_cap  = t->capacity;
    slot *new_slots = (slot *)t->allocator->alloc(t->allocator, new_capacity * sizeof(slot), alignof(slot));
    if (!new_slots) return KE_ERROR_OUT_OF_MEMORY;
    memset(new_slots, 0, new_capacity * sizeof(slot));

    t->slots = new_slots;
    t->capacity = new_capacity;
    t->occupied = 0;
    t->active = 0;

    for (size_t i = 0; i < old_cap; i++) {
        const slot *s = &old_slots[i];
        if (s->key == RC_EMPTY || s->key == RC_TOMBSTONE) continue;
        bool found = false;
        size_t idx = table_probe(t, s->key, &found);
        t->slots[idx] = *s;
        t->occupied++;
        t->active++;
    }
    t->allocator->free(t->allocator, old_slots);
    return KE_OK;
}

static uint64_t key_from_handle(ke_resource_handle h) {
    return (uint64_t)h + 1ULL;  // +1 keeps 0 reserved for EMPTY
}

static uint64_t key_from_path(const char *path) {
    uint64_t k = ke_hash_string(path);
    if (k == RC_EMPTY)     return 1ULL;
    if (k == RC_TOMBSTONE) return RC_TOMBSTONE - 1ULL;
    return k;
}

// ── vtable: lifetime ────────────────────────────────────────────────────────

static ke_result vt_register(ke_resource_cache *self, ke_resource_handle handle) {
    if (!self || handle == KE_RESOURCE_HANDLE_NONE) return KE_ERROR_INVALID_ARGUMENT;
    rc_state *s = (rc_state *)self->handle;

    ke_result rc = table_reserve(&s->resources);
    if (rc != KE_OK) return rc;

    bool found = false;
    uint64_t key = key_from_handle(handle);
    size_t idx = table_probe(&s->resources, key, &found);
    if (found) return KE_ERROR_INVALID_ARGUMENT;  // double-register

    if (s->resources.slots[idx].key == RC_EMPTY) s->resources.occupied++;
    s->resources.active++;
    s->resources.slots[idx].key      = key;
    s->resources.slots[idx].refcount = 1;
    s->resources.slots[idx].handle   = handle;
    return KE_OK;
}

static ke_result vt_retain(ke_resource_cache *self, ke_resource_handle handle) {
    if (!self || handle == KE_RESOURCE_HANDLE_NONE) return KE_ERROR_INVALID_ARGUMENT;
    rc_state *s = (rc_state *)self->handle;
    bool found = false;
    size_t idx = table_probe(&s->resources, key_from_handle(handle), &found);
    if (!found) return KE_ERROR_NOT_FOUND;
    s->resources.slots[idx].refcount++;
    return KE_OK;
}

// Walks the path table and tombstones any entry pointing at `handle`. Called
// from release() when refcount hits zero so dedup cache stays consistent.
static void evict_paths_for_handle(rc_state *s, ke_resource_handle handle) {
    for (size_t i = 0; i < s->paths.capacity; i++) {
        slot *p = &s->paths.slots[i];
        if (p->key != RC_EMPTY && p->key != RC_TOMBSTONE && p->handle == handle) {
            p->key = RC_TOMBSTONE;
            s->paths.active--;
        }
    }
}

static ke_result vt_release(ke_resource_cache *self, ke_resource_handle handle) {
    if (!self || handle == KE_RESOURCE_HANDLE_NONE) return KE_ERROR_INVALID_ARGUMENT;
    rc_state *s = (rc_state *)self->handle;
    bool found = false;
    size_t idx = table_probe(&s->resources, key_from_handle(handle), &found);
    if (!found) return KE_ERROR_NOT_FOUND;

    slot *slot_ref = &s->resources.slots[idx];
    if (slot_ref->refcount == 0) return KE_ERROR_INVALID_ARGUMENT;
    slot_ref->refcount--;
    if (slot_ref->refcount > 0) return KE_OK;

    slot_ref->key = RC_TOMBSTONE;
    slot_ref->refcount = 0;
    s->resources.active--;

    evict_paths_for_handle(s, handle);

    if (s->destroy_fn) s->destroy_fn(handle, s->destroy_ctx);
    return KE_OK;
}

// ── vtable: path cache ──────────────────────────────────────────────────────

static bool vt_try_get_cached(ke_resource_cache *self, const char *key, ke_resource_handle *out_handle) {
    if (!self || !key || !out_handle) return false;
    rc_state *s = (rc_state *)self->handle;
    bool found = false;
    size_t idx = table_probe(&s->paths, key_from_path(key), &found);
    if (!found) return false;

    ke_resource_handle h = s->paths.slots[idx].handle;
    // Retain on behalf of caller. If the underlying resource vanished
    // (stale path entry), drop the path entry and report cache miss.
    if (vt_retain(self, h) != KE_OK) {
        s->paths.slots[idx].key = RC_TOMBSTONE;
        s->paths.active--;
        return false;
    }
    *out_handle = h;
    return true;
}

static ke_result vt_cache_insert(ke_resource_cache *self, const char *key, ke_resource_handle handle) {
    if (!self || !key || handle == KE_RESOURCE_HANDLE_NONE) return KE_ERROR_INVALID_ARGUMENT;
    rc_state *s = (rc_state *)self->handle;
    ke_result rc = table_reserve(&s->paths);
    if (rc != KE_OK) return rc;

    bool found = false;
    uint64_t k = key_from_path(key);
    size_t idx = table_probe(&s->paths, k, &found);
    if (found) return KE_ERROR_INVALID_ARGUMENT;  // duplicate key — caller must try_get_cached first

    if (s->paths.slots[idx].key == RC_EMPTY) s->paths.occupied++;
    s->paths.active++;
    s->paths.slots[idx].key      = k;
    s->paths.slots[idx].refcount = 0;
    s->paths.slots[idx].handle   = handle;
    return KE_OK;
}

static void vt_cache_evict(ke_resource_cache *self, const char *key) {
    if (!self || !key) return;
    rc_state *s = (rc_state *)self->handle;
    bool found = false;
    size_t idx = table_probe(&s->paths, key_from_path(key), &found);
    if (!found) return;
    s->paths.slots[idx].key = RC_TOMBSTONE;
    s->paths.active--;
}

// ── vtable: teardown ────────────────────────────────────────────────────────

static void vt_destroy(ke_resource_cache *self) {
    if (!self) return;
    rc_state *s = (rc_state *)self->handle;
    if (!s) return;

    // Fire destroy callback for every resource still alive.
    for (size_t i = 0; i < s->resources.capacity; i++) {
        slot *r = &s->resources.slots[i];
        if (r->key == RC_EMPTY || r->key == RC_TOMBSTONE) continue;
        ke_resource_handle h = r->handle;
        r->key = RC_TOMBSTONE;
        if (s->destroy_fn) s->destroy_fn(h, s->destroy_ctx);
    }

    table_destroy(&s->resources);
    table_destroy(&s->paths);
    ke_allocator *a = s->allocator;
    a->free(a, s);
    a->destroy(a);
}

// ── Factory ─────────────────────────────────────────────────────────────────

ke_result ke_resource_cache_create(const ke_resource_cache_params *params,
                                    ke_resource_cache             **out_cache) {
    if (!params || !out_cache) return KE_ERROR_INVALID_ARGUMENT;

    ke_allocator *a = ke_allocator_malloc_create();
    if (!a) return KE_ERROR_OUT_OF_MEMORY;

    rc_state *s = (rc_state *)a->alloc(a, sizeof(rc_state), alignof(rc_state));
    if (!s) { a->destroy(a); return KE_ERROR_OUT_OF_MEMORY; }
    memset(s, 0, sizeof(*s));

    s->allocator   = a;
    s->destroy_fn  = params ? params->destroy_fn  : NULL;
    s->destroy_ctx = params ? params->destroy_ctx : NULL;

    ke_result rc = table_init(&s->resources, a, 64);
    if (rc != KE_OK) { a->free(a, s); a->destroy(a); return rc; }
    rc = table_init(&s->paths, a, 64);
    if (rc != KE_OK) { table_destroy(&s->resources); a->free(a, s); a->destroy(a); return rc; }

    s->api.handle            = s;
    s->api.register_resource = vt_register;
    s->api.retain            = vt_retain;
    s->api.release           = vt_release;
    s->api.try_get_cached    = vt_try_get_cached;
    s->api.cache_insert      = vt_cache_insert;
    s->api.cache_evict       = vt_cache_evict;
    s->api.destroy           = vt_destroy;

    *out_cache = &s->api;
    return KE_OK;
}

File: src/zig/resource_cache/default/src/resource_cache.zig
const std = @import("std");

const gpa = std.heap.c_allocator;

const c = @cImport({
    @cInclude("kernel_engine/resource_cache/resource_cache.h");
});

// Zig-native error translation at the C-ABI seam (no ke_common link).
const E = @import("kerror").Errors(c);

// Open-addressed table with tombstones for both the handle→entry and
// path→handle directions. destroy_fn is per-cache (set at construction); all
// resources in a cache share the same destructor.

fn hashString(str: ?[*:0]const u8) u64 {
    const s = str orelse return 0;
    var h: u64 = 5381;
    var i: usize = 0;
    while (s[i] != 0) : (i += 1) {
        h = h *% 33 ^ @as(u64, s[i]);
    }
    return h;
}

const RC_EMPTY: u64 = 0;
const RC_TOMBSTONE: u64 = std.math.maxInt(u64);
const HANDLE_NONE: c.ke_resource_handle = std.math.maxInt(u32);

const Slot = struct {
    key: u64 = 0, // 0 = empty, maxU64 = tombstone
    refcount: u32 = 0, // resource entries only
    handle: u32 = 0, // resource: redundant; path: target handle
};

const Table = struct {
    slots: []Slot,
    capacity: usize, // power of two
    occupied: usize, // active + tombstones
    active: usize, // active only

    fn init(initial_capacity: usize) ?Table {
        const slots = gpa.alloc(Slot, initial_capacity) catch return null;
        @memset(slots, .{});
        return .{ .slots = slots, .capacity = initial_capacity, .occupied = 0, .active = 0 };
    }

    fn deinit(t: *Table) void {
        if (t.slots.len != 0) gpa.free(t.slots);
        t.slots = &.{};
        t.capacity = 0;
        t.occupied = 0;
        t.active = 0;
    }

    // Finds the slot for `key`. Returns the matching active entry, or the first
    // usable slot (preferring tombstones over empty so reinsertion keeps probe
    // chains short). `found` discriminates.
    fn probe(t: *const Table, key: u64, found: *bool) usize {
        const mask = t.capacity - 1;
        var index = @as(usize, @intCast(key & mask));
        var first_tombstone: usize = std.math.maxInt(usize);
        var i: usize = 0;
        while (i < t.capacity) : (i += 1) {
            const s = &t.slots[index];
            if (s.key == RC_EMPTY) {
                found.* = false;
                return if (first_tombstone != std.math.maxInt(usize)) first_tombstone else index;
            }
            if (s.key == RC_TOMBSTONE) {
                if (first_tombstone == std.math.maxInt(usize)) first_tombstone = index;
            } else if (s.key == key) {
                found.* = true;
                return index;
            }
            index = (index + 1) & mask;
        }
        found.* = false;
        return if (first_tombstone != std.math.maxInt(usize)) first_tombstone else 0;
    }

    fn reserve(t: *Table) bool {
        if ((t.occupied + 1) * 10 < t.capacity * 7) return true;
        return t.rehash(t.capacity * 2);
    }

    fn rehash(t: *Table, new_capacity: usize) bool {
        const old_slots = t.slots;
        const new_slots = gpa.alloc(Slot, new_capacity) catch return false;
        @memset(new_slots, .{});

        t.slots = new_slots;
        t.capacity = new_capacity;
        t.occupied = 0;
        t.active = 0;

        for (old_slots) |s| {
            if (s.key == RC_EMPTY or s.key == RC_TOMBSTONE) continue;
            var found = false;
            const idx = t.probe(s.key, &found);
            t.slots[idx] = s;
            t.occupied += 1;
            t.active += 1;
        }
        gpa.free(old_slots);
        return true;
    }
};

fn keyFromHandle(h: c.ke_resource_handle) u64 {
    return @as(u64, h) + 1; // +1 keeps 0 reserved for EMPTY
}

fn keyFromPath(path: ?[*:0]const u8) u64 {
    const k = hashString(path);
    if (k == RC_EMPTY) return 1;
    if (k == RC_TOMBSTONE) return RC_TOMBSTONE - 1;
    return k;
}

const State = struct {
    api: c.ke_resource_cache,
    destroy_fn: c.ke_resource_destroy_func,
    destroy_ctx: ?*anyopaque,
    resources: Table,
    paths: Table,
};

fn stateOf(self: *c.ke_resource_cache) *State {
    return @ptrCast(@alignCast(self.handle));
}

// ── vtable: lifetime ────────────────────────────────────────────────────────

fn vtRegister(self: ?*c.ke_resource_cache, handle: c.ke_resource_handle, out_error: [*c][*c]c.ke_error) callconv(.c) bool {
    if (self == null or handle == HANDLE_NONE) {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return false;
    }
    const s = stateOf(self.?);
    if (!s.resources.reserve()) return false;

    var found = false;
    const key = keyFromHandle(handle);
    const idx = s.resources.probe(key, &found);
    if (found) {
        E.fail(out_error, .already_exists, "handle already registered", @src());
        return false;
    }

    if (s.resources.slots[idx].key == RC_EMPTY) s.resources.occupied += 1;
    s.resources.active += 1;
    s.resources.slots[idx] = .{ .key = key, .refcount = 1, .handle = handle };
    return true;
}

fn vtRetain(self: ?*c.ke_resource_cache, handle: c.ke_resource_handle, out_error: [*c][*c]c.ke_error) callconv(.c) bool {
    if (self == null or handle == HANDLE_NONE) {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return false;
    }
    const s = stateOf(self.?);
    var found = false;
    const idx = s.resources.probe(keyFromHandle(handle), &found);
    if (!found) {
        E.fail(out_error, .not_found, "handle not found", @src());
        return false;
    }
    s.resources.slots[idx].refcount += 1;
    return true;
}

// Walks the path table and tombstones any entry pointing at `handle`.
fn evictPathsForHandle(s: *State, handle: c.ke_resource_handle) void {
    for (s.paths.slots) |*p| {
        if (p.key != RC_EMPTY and p.key != RC_TOMBSTONE and p.handle == handle) {
            p.key = RC_TOMBSTONE;
            s.paths.active -= 1;
        }
    }
}

fn vtRelease(self: ?*c.ke_resource_cache, handle: c.ke_resource_handle, out_error: [*c][*c]c.ke_error) callconv(.c) bool {
    if (self == null or handle == HANDLE_NONE) {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return false;
    }
    const s = stateOf(self.?);
    var found = false;
    const idx = s.resources.probe(keyFromHandle(handle), &found);
    if (!found) {
        E.fail(out_error, .not_found, "handle not found", @src());
        return false;
    }

    const slot_ref = &s.resources.slots[idx];
    if (slot_ref.refcount == 0) {
        E.fail(out_error, .general, "refcount is already zero", @src());
        return false;
    }
    slot_ref.refcount -= 1;
    if (slot_ref.refcount > 0) return true;

    slot_ref.key = RC_TOMBSTONE;
    slot_ref.refcount = 0;
    s.resources.active -= 1;

    evictPathsForHandle(s, handle);

    if (s.destroy_fn) |d| d(handle, s.destroy_ctx);
    return true;
}

// ── vtable: path cache ──────────────────────────────────────────────────────

fn vtTryGetCached(self: ?*c.ke_resource_cache, key: [*c]const u8, out_handle: [*c]c.ke_resource_handle) callconv(.c) bool {
    if (self == null or key == null or out_handle == null) return false;
    const s = stateOf(self.?);
    var found = false;
    const idx = s.paths.probe(keyFromPath(key), &found);
    if (!found) return false;

    const h = s.paths.slots[idx].handle;
    // Retain on behalf of caller. If the underlying resource vanished (stale
    // path entry), drop the path entry and report cache miss.
    if (!vtRetain(self, h, null)) {
        s.paths.slots[idx].key = RC_TOMBSTONE;
        s.paths.active -= 1;
        return false;
    }
    out_handle.* = h;
    return true;
}

fn vtCacheInsert(self: ?*c.ke_resource_cache, key: [*c]const u8, handle: c.ke_resource_handle, out_error: [*c][*c]c.ke_error) callconv(.c) bool {
    if (self == null or key == null or handle == HANDLE_NONE) {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return false;
    }
    const s = stateOf(self.?);
    if (!s.paths.reserve()) {
        E.fail(out_error, .general, "path table rehash failed", @src());
        return false;
    }

    var found = false;
    const k = keyFromPath(key);
    const idx = s.paths.probe(k, &found);
    if (found) {
        E.fail(out_error, .already_exists, "key already in cache; call try_get_cached first", @src());
        return false;
    }

    if (s.paths.slots[idx].key == RC_EMPTY) s.paths.occupied += 1;
    s.paths.active += 1;
    s.paths.slots[idx] = .{ .key = k, .refcount = 0, .handle = handle };
    return true;
}

fn vtCacheEvict(self: ?*c.ke_resource_cache, key: [*c]const u8) callconv(.c) void {
    if (self == null or key == null) return;
    const s = stateOf(self.?);
    var found = false;
    const idx = s.paths.probe(keyFromPath(key), &found);
    if (!found) return;
    s.paths.slots[idx].key = RC_TOMBSTONE;
    s.paths.active -= 1;
}

// ── vtable: teardown ────────────────────────────────────────────────────────

fn vtDestroy(self: ?*c.ke_resource_cache) callconv(.c) void {
    const api = self orelse return;
    const s = stateOf(api);

    // Fire destroy callback for every resource still alive.
    for (s.resources.slots) |*r| {
        if (r.key == RC_EMPTY or r.key == RC_TOMBSTONE) continue;
        const h = r.handle;
        r.key = RC_TOMBSTONE;
        if (s.destroy_fn) |d| d(h, s.destroy_ctx);
    }

    s.resources.deinit();
    s.paths.deinit();
    gpa.destroy(s);
}

// ── Factory ─────────────────────────────────────────────────────────────────

export fn ke_resource_cache_create(params: [*c]const c.ke_resource_cache_params, out_error: [*c][*c]c.ke_error) callconv(.c) c.ke_resource_cache_handle {
    const empty = c.ke_resource_cache_handle{ .ref = null, .destroy = null };
    if (params == null) {
        E.fail(out_error, .invalid_argument, "invalid argument", @src());
        return empty;
    }

    const s = gpa.create(State) catch {
        E.fail(out_error, .out_of_memory, "state allocation failed", @src());
        return empty;
    };
    s.* = std.mem.zeroes(State);
    s.destroy_fn = params.*.destroy_fn;
    s.destroy_ctx = params.*.destroy_ctx;

    s.resources = Table.init(64) orelse {
        gpa.destroy(s);
        E.fail(out_error, .out_of_memory, "resources table allocation failed", @src());
        return empty;
    };
    s.paths = Table.init(64) orelse {
        s.resources.deinit();
        gpa.destroy(s);
        E.fail(out_error, .out_of_memory, "paths table allocation failed", @src());
        return empty;
    };

    s.api.handle = s;
    s.api.register_resource = &vtRegister;
    s.api.retain = &vtRetain;
    s.api.release = &vtRelease;
    s.api.try_get_cached = &vtTryGetCached;
    s.api.cache_insert = &vtCacheInsert;
    s.api.cache_evict = &vtCacheEvict;

    return .{ .ref = &s.api, .destroy = &vtDestroy };
}
